### ISSUE TITLE
Easy way to integrate the paranoia gem?

### ISSUE THREAD

**Issue by mikeyhogarth** [toptal/chewy#326]
We have a requirement to integrate the paranoia gem into our system:
https://github.com/rubysherpas/paranoia

Which basically provides 'soft delete' for records (sets a flag of 'deleted' on them rather than full on destroying them) - the problem is that the paranoia gem hooks into the model callbacks so they're being removed from the ES index and we need them to stay in there!

I've had a quick dig around and it doesn't look like there's a way of skipping this at all - I am suspecting that some kind of monkey patch or extension is going to be required but before I go down that route I wondered if this had come up before and what the solution was?

Basically all we need is a way of going: 

* Before deleting from the ES index, make sure it actually HAS been properly deleted.

**Comment by krzysiek1507** [comment]
@mikeyhogarth just add scope `with_deleted` in your type in index class.
The minus is reindexing of this record.
```ruby
  define_type Spree::Product.with_deleted do
    ...
  end
```

**Comment by mikeyhogarth** [comment]
Thank you so much!

**Comment by krzysiek1507** [comment]
This solution is problematic because Chewy caches this query, so when you have query based on time, then this time is the same for all saves @pyromaniac

**Comment by pyromaniac** [comment]
Didn't get what do you mean

**Comment by krzysiek1507** [comment]
```ruby
class Spree::Product
  update_index 'spree/products#product', :self
end

class Spree::ProductsIndex < Chewy::Index
  define_type Spree::Product.where(created_at: Time.zone.now) do
    field :name
  end
end
```

First save
```ruby
puts Time.zone.now
# 2016-03-30 14:07:59
Spree::Product.last.save

# Chewy's output
SELECT COUNT(*) FROM "spree_products" WHERE "spree_products"."created_at" = $1 AND "spree_products"."id" IN (6949)  [["created_at", "2016-03-30 12:07:59.403380"]]
Spree::ProductsIndex::Product Import (6.7ms) {:delete=>1}
```

Second save
```ruby
puts Time.zone.now
# 2016-03-30 14:09:03
Spree::Product.last.save

# Chewy's output
SELECT COUNT(*) FROM "spree_products" WHERE "spree_products"."created_at" = $1 AND "spree_products"."id" IN (6949)  [["created_at", "2016-03-30 12:07:59.403380"]]
Spree::ProductsIndex::Product Import (6.6ms) {:delete=>1}
```

Take a look on time and `created_at` in queries

**Comment by pyromaniac** [comment]
Ah, right, this is by design, and this is not related to paranoid actually. I'm not sure what do you want to achieve with this btw, you probably don't want this. You want to use just `define_type Spree::Product do` and everything will work just fine for you.

**Comment by krzysiek1507** [comment]
So how to use it with this scope and `update_index` callback?
```ruby
scope :last_week, -> { where('available_on BETWEEN :start AND :end', start: 7.days.ago, end: Time.zone.now) }
```

**Comment by pyromaniac** [comment]
Ah, so you want index only for products from the previous week, right? Would not it be better to index everything, but restrict dates on query level already?

**Comment by krzysiek1507** [comment]
Right. But what about callback on product save? I can change `available_on` manually to `14.days.ago` or `2.days.from_now`.

**Comment by pyromaniac** [comment]
Yep, and it will be simply reindexed

**Comment by krzysiek1507** [comment]
So how to use it?

**Comment by pyromaniac** [comment]
Spree::Product.update_attributes(available_on: 14.days.ago)

**Comment by krzysiek1507** [comment]
Indexes all (wrong):
```ruby
class Spree::Product
  update_index 'spree/products#product', :self
end

class Spree::ProductsIndex < Chewy::Index
  define_type Spree::Product do
    field :name
  end
end
```

Wrong query because of cached value of `Time.zone.now`:
```ruby
class Spree::Product
  update_index 'spree/products#product', :self
end

class Spree::ProductsIndex < Chewy::Index
  define_type Spree::Product.last_week do
    field :name
  end
end
```

Third one?

**Comment by pyromaniac** [comment]
I still don't understand what are you trying to achieve and why the first option is wrong?

**Comment by krzysiek1507** [comment]
First option omit conditions, so `update_index` callback will index products not in `last_week` scope too.
`Spree::ProductsIndex.import` will index all products. I don't want to add everywhere this scope, because then I have to tell everyone to use it and I have to remember about it.
Maybe develop option `scope` for `define_type` (like in `load`)?

**Comment by pyromaniac** [comment]
You can actually use `define_type -> { scope } do`, but it will use object adapter instead of active_record, so I have no idea which problems may it cause. Also seems to me that you will abandon this idea anyway because it doesn't look reasonable and you are overcomplicating things right now.

**Comment by pyromaniac** [comment]
Hey, @krzysiek1507 why don't you calculate time now in SQL instead of ruby?

**Comment by krzysiek1507** [comment]
@pyromaniac that scope was just example. Scopes are from gem and I don't want to override.
